piksi_multi_cpp/src/receiver.cc

```diff
@@ -0,0 +1,131 @@
+#include "piksi_multi_cpp/receiver.h"
+
+#include <libsbp/sbp.h>
+#include <piksi_rtk_msgs/Heartbeat.h>
+#include "piksi_multi_cpp/device.h"
+
+// Forward declarations
+#include "piksi_multi_cpp/receiver_attitude.h"
+#include "piksi_multi_cpp/receiver_base_station.h"
+#include "piksi_multi_cpp/receiver_position.h"
+
+// SBP message definitions.
+#include <libsbp/system.h>
+
+namespace piksi_multi_cpp {
+
+std::vector<Receiver::Type> Receiver::kTypeVec = std::vector<Receiver::Type>(
+    {kBaseStationReceiver, kPositionReceiver, kAttitudeReceiver, kUnknown});
+
+Receiver::Receiver(const ros::NodeHandle& nh,
+                   const std::shared_ptr<Device>& device)
+    : nh_(nh), device_(device) {
+  // Initialize SBP state.
+  state_ = std::make_shared<sbp_state_t>();
+  sbp_state_init(state_.get());
+
+  // Register callbacks.
+  cb_.push_back(Callback::create(nh, SBP_MSG_HEARTBEAT, state_));
+}
+
+std::shared_ptr<Receiver> Receiver::create(
+    const ros::NodeHandle& nh, const std::shared_ptr<Device>& device,
+    const Type type) {
+  switch (type) {
+    case Type::kBaseStationReceiver:
+      return std::shared_ptr<Receiver>(new ReceiverBaseStation(nh, device));
```

Comment:
Hm, this creates a sort of circular dependency, as Receiver needs to know all its child types, and the child types need to know Receiver for inheritance